src/functions.php

```diff
@@ -94,15 +95,34 @@ function apply_type_map_to_document($document, array $typeMap)
 }
 
 /**
- * Generate an index name from a key specification.
+ * Converts a document parameter to an array.
+ *
+ * This is used to facilitate unified access to document fields. It also handles
+ * Document, PackedArray, and Serializable objects.
+ *
+ * PackedArray is intentionally allowed because this function is not used for
+ * type checking. Prohibiting PackedArray would only require callers to check
+ * for a PackedArray in order to avoid an exception. Furthermore, this function
+ * does not distinguish between Serializable::bsonSerialize() return values that
+ * might encode as a BSON document or array.
  *
  * @internal
- * @param array|object $document Document containing fields mapped to values,
- *                               which denote order or an index type
- * @throws InvalidArgumentException
+ * @param array|object $document
```

Comment:
The phpdoc is about what is expected. It's not a validation function that should accept anything and fail in case of invalid input. We need to throw a `TypeError`. When upgrading to PHP 8.0+ the union type will move to the code.

Reply:
> The phpdoc is about what is expected.

Correct. Only arrays or objects should be provided to `document_to_array()`.

> We need to throw a TypeError.

PHPLIB's own type checking uses `InvalidArgumentException::invalidType()` everywhere. I suggest keeping that here for consistency. When we can rely on PHP itself to enforce the parameter type, then we'll get TypeError.

That said, both of these exceptions should be very unlikely for the following reasons:

 * Excluding test cases, every code path that invokes `document_to_array()` should already have validated the `$document` before doing so.
 * The edge case for `document_to_array()` being passed an object and still throwing has to do with `Serializable::bsonSerialize()` returning a non-array, non-object. That should only be possible on PHP 7.x, since [8.0+ declares a union return type hint](https://github.com/mongodb/mongo-php-driver/blob/1.15.3/src/BSON/Serializable.stub.php#L12-L18).

Reply:
Ah, I see. I saw this as a validation function which would take anything. I can't wait until we can bump the PHP requirement to PHP 8.2 (or at least 8.1 to get intersection types) and make full use of the PHP type system.